hs/src/Reach/Type.hs

```diff
@@ -247,29 +249,30 @@ typeChecks tint@(mcfs, _) at env ts vs =
     ([], []) ->
       return []
     ((t : ts'), (v : vs')) -> do
-      d <- typeCheck tint at env t v
+      d <- typeCheck tint at env (dt2st t) v
       ds' <- typeChecks tint at env ts' vs'
       return $ d : ds'
     ((_ : _), _) ->
       expect_throw mcfs at $ Err_Type_TooFewArguments ts
     (_, (_ : _)) ->
       expect_throw mcfs at $ Err_Type_TooManyArguments vs
 
-checkAndConvert_i :: TINT -> SrcLoc -> TypeEnv s -> DLType -> [SLVal] -> ST s (DLType, [DLArgExpr])
+checkAndConvert_i :: TINT -> SrcLoc -> TypeEnv s -> SLType -> [SLVal] -> ST s (DLType, [DLArgExpr])
 checkAndConvert_i tint@(mcfs, _) at env t args =
   case t of
-    T_Fun dom rng -> do
-      dargs <- typeChecks tint at env dom args
-      return (rng, dargs)
-    T_Forall var ft -> do
+    ST_Fun dom rng -> do
+      dargs <- typeChecks tint at env (map st2dt dom) args
+      return (st2dt rng, dargs)
+    ST_Forall var ft -> do
       var_ref <- newSTRef Nothing
       let env' = M.insert var var_ref env
-      (vrng, dargs) <- checkAndConvert_i tint at env' ft args
-      rng <- typeSubst at env' vrng
+      (rng, dargs) <- checkAndConvert_i tint at env' ft args
+      -- XXX happens prior to this point now?
+      -- rng <- typeSubst at env' vrng
```

Comment:
So, if the types preclude that... then this suggests there might be another place where a `st2dt` would error because of the type variable not having been substituted